llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
+  // form a back-reference if we loop back here in future.
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, GODepth);
   if (!Inserted) {
-    // If GO is visited, use the letter 'R' as the marker and use its state as
-    // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    assert(GODepth > StateIt->second);
+    // If GO is visited, mark as a Backref and use its state as the value.
+    GOHash.update(static_cast<char>(Tags::Backref));
+    GOHash.update(GODepth - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    DigestType Digest;
+    GOHash.final(Digest);
+    Changed = true;
+    return std::make_tuple(StateIt->second, Digest);
   }
 
-  GOHash.update('T');
-  auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  bool Changed = State.Changed;
+  GOHash.update(static_cast<char>(Tags::GO));
+  auto State = accumulateGODigest(GO, GOHash, HasRepoDefinitionMD);
   // The hashes of the GO's 'contributions' and 'dependencies' are separately
   // added to the GO's hash since their respective arrows point are in different
   // direction. In addition, 'contributions' cannot form loops since 1) only
   // global variables have 'contributions' and 2) all 'contributions' are
   // functions. Therefore, the 'contributions' can be simplified not to record
   // 'visited' table, which avoids having two 'visited' maps. Firstly, add the
   // GO's contributions to its hash.
-  Changed = updateDigestUseContributions(GOHash, State.Contributions, GOIMap) ||
-            Changed;
+  updateDigestUseContributions(GOHash, State.Contributions);
 
   // Add the GO's dependencies to its hash.
-  Changed = updateDigestUseDependencies(GOHash, State.Dependencies, Visited,
-                                        GOIMap, AccumulateGODigest) ||
-            Changed;
-
-  return Changed;
+  const auto result = updateDigestUseDependencies(
+      GO, GOHash, GODepth, State.Dependencies, HasRepoDefinitionMD);
+  Visited.erase(GO);
+  LLVM_DEBUG(dbgs() << "Removing \"" << GO->getName()
+                    << "\" from the Visited map\n");
+  return result;
 }
 
-// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M) {
-  GOInfoMap GOIMap;
-  GONumber GONum;
-
-  for (auto &GV : M.globals()) {
-    if (GV.isDeclaration())
-      continue;
-    calculateGOInfo(&GV, GOIMap);
-    ++GONum.VarNum;
-  }
-  for (auto &Fn : M.functions()) {
-    if (Fn.isDeclaration())
-      continue;
-    calculateGOInfo(&Fn, GOIMap);
-    ++GONum.FuncNum;
-  }
-  return std::make_pair(std::move(GOIMap), std::move(GONum));
+DigestType ModuleHashGenerator::calculateDigest(const GlobalObject *GO,
+                                                bool HasRepoDefinitionMD) {
+  auto const Result = std::get<DigestType>(
```

Comment:
Given that Visited is a member of this class, it could also be asserted that it is empty before starting to use it...

Reply:
Added, thanks.